core/offchain/impl/offchain_worker_impl.cpp

```diff
@@ -0,0 +1,340 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "offchain/impl/offchain_worker_impl.hpp"
+
+#include <libp2p/host/host.hpp>
+#include <thread>
+
+#include "api/service/author/author_api.hpp"
+#include "application/app_configuration.hpp"
+#include "crypto/hasher.hpp"
+#include "offchain/impl/offchain_local_storage.hpp"
+#include "runtime/executor.hpp"
+
+namespace kagome::offchain {
+
+  OffchainWorkerImpl::OffchainWorkerImpl(
+      const application::AppConfiguration &app_config,
+      std::shared_ptr<clock::SystemClock> clock,
+      std::shared_ptr<crypto::Hasher> hasher,
+      std::shared_ptr<storage::BufferStorage> storage,
+      std::shared_ptr<crypto::CSPRNG> random_generator,
+      std::shared_ptr<api::AuthorApi> author_api,
+      libp2p::Host &host,
+      std::shared_ptr<OffchainPersistentStorage> persistent_storage,
+      std::shared_ptr<runtime::Executor> executor,
+      const primitives::BlockHeader &header)
+      : app_config_(app_config),
+        clock_(std::move(clock)),
+        hasher_(std::move(hasher)),
+        random_generator_(std::move(random_generator)),
+        author_api_(std::move(author_api)),
+        host_(host),
+        persistent_storage_(std::move(persistent_storage)),
+        executor_(std::move(executor)),
+        header_(header),
+        log_(log::createLogger(
+            "OffchainWorker#" + std::to_string(header_.number), "offchain")) {
+    BOOST_ASSERT(clock_);
+    BOOST_ASSERT(hasher_);
+    BOOST_ASSERT(storage);
+    BOOST_ASSERT(random_generator_);
+    BOOST_ASSERT(author_api_);
+    BOOST_ASSERT(persistent_storage_);
+    BOOST_ASSERT(executor_);
+
+    auto hash = hasher_->blake2b_256(scale::encode(header_).value());
```

Comment:
I think we should have a more centralised way of obtaining a block hash, I mildly dislike that the information that it's blake2b from the SCALE encoding of the header leaks everywhere we need a block hash.

Reply:
Good idea! 
I made new issue for that: #999